### PR TITLE
Further improvements to CSX support

### PR DESCRIPTION
**Pull request by filipw**
This is a follow up PR to the large work we did in https://github.com/OmniSharp/omnisharp-roslyn/pull/659
In that PR we removed custom scripting code and focused on "standard" Roslyn scripting.

In this PR, we take it a step further and introduce suggestions from Tomas (https://github.com/OmniSharp/omnisharp-roslyn/pull/659#issuecomment-259746769) on how to approach CSX projects (which have no project system after all). So what we do here, is we move away from the current model of "project-per-file" to a "project-per-entry-file" model (a project is created only for a top-level CSX file and all its #load dependencies are delegated to Roslyn's `ScriptSourceResolver` - we never manually specify project-to-project relationships anymore).

This has some terrific results for us:
 - we can **delete a ton** of custom parsing code, and let Roslyn resolvers do the heavy lifting for us
 - we get support for workspace level changes out of the box - user can change `#r` references or `#load` directives and we pick them up in real time
 - multiple `#load` per project are now supported (in the past, due to our  "project-per-file" model, only 1 was allowed - this is because a submission compilation can only have one project reference)

Fixes https://github.com/OmniSharp/omnisharp-roslyn/issues/227 and https://github.com/OmniSharp/omnisharp-roslyn/issues/689 - two most important issues we have for CSX at the moment.

Here is the behavior demo for "real time" `#load`:

And "real time" `#r`:

### CODE REVIEW

File: src/OmniSharp.Roslyn.CSharp/Services/Navigation/GotoDefinitionService.cs [hunk 1]
```diff
@@ -55,25 +55,32 @@ public GotoDefinitionService(OmniSharpWorkspace workspace, MetadataHelper metada
                     }
                     else if (location.IsInMetadata && request.WantMetadata)
                     {
-                        var cancellationSource = new CancellationTokenSource(TimeSpan.FromMilliseconds(request.Timeout));
-                        var metadataDocument = await _metadataHelper.GetDocumentFromMetadata(document.Project, symbol, cancellationSource.Token);
-                        if (metadataDocument != null)
+                        try
```
Comment (filipw) [inline src/OmniSharp.Roslyn.CSharp/Services/Navigation/GotoDefinitionService.cs]: sorry this change got in by accident, will revert

File: src/OmniSharp.Roslyn/MetadataHelper.cs [hunk 2]
```diff
@@ -62,8 +62,7 @@ public Task<Document> GetDocumentFromMetadata(Project project, ISymbol symbol, C
 
             var service = _csharpMetadataAsSourceService.CreateInstance(temporaryDocument.Project.LanguageServices);
             var method = _csharpMetadataAsSourceService.GetMethod(AddSourceToAsync);
-
-            return method.Invoke<Task<Document>>(service, new object[] { temporaryDocument, topLevelSymbol, cancellationToken });
+                return method.Invoke<Task<Document>>(service, new object[] {temporaryDocument, topLevelSymbol, cancellationToken});
```
Comment (DustinCampbell) [inline src/OmniSharp.Roslyn/MetadataHelper.cs]: Did you mean to indent here?

File: src/OmniSharp.Script/ScriptProjectSystem.cs [hunk 3]
```diff
@@ -107,28 +131,43 @@ public void Initalize(IConfiguration configuration)
...
+            // Each .CSX file becomes an entry point for it's own project
+            // Every #loaded file will be part of the project too
             foreach (var csxPath in allCsxFiles)
             {
                 try
                 {
-                    CreateCsxProject(csxPath);
+                    var csxFileName = Path.GetFileName(csxPath);
+                    var project = ProjectInfo.Create(
+                        id: ProjectId.CreateNewId(Guid.NewGuid().ToString()),
```
Comment (DustinCampbell) [inline src/OmniSharp.Script/ScriptProjectSystem.cs]: I don't think you need to create a new Guid for this. `ProjectId.Create()` already does that. The `debugName` parameter is to give it something human-readable for debugging. ([ref](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/ProjectId.cs,bac27363e2d822dc))

File: src/OmniSharp.Script/ScriptProjectSystem.cs [hunk 4]
```diff
@@ -107,28 +131,43 @@ public void Initalize(IConfiguration configuration)
...
+                    _workspace.AddDocument(project.Id, csxPath, SourceCodeKind.Script);
+                    _projects[csxPath] = project;
+                    _logger.LogInformation($"Added CSX project '{csxPath}' to the workspace.");
                 }
                 catch (Exception ex)
                 {
-                    Logger.LogError($"{csxPath} will be ignored due to the following error:", ex.ToString());
-                    Logger.LogError(ex.ToString());
-                    Logger.LogError(ex.InnerException?.ToString() ?? "No inner exception.");
+                    _logger.LogError(0, ex, $"{csxPath} will be ignored due to an following error");
```
Comment (DustinCampbell) [inline src/OmniSharp.Script/ScriptProjectSystem.cs]: We've got an extension method that allows you to call LogError without needing to pass an event ID now.

File: src/OmniSharp.Roslyn/MetadataHelper.cs (hunk 2 above)
Comment (filipw) [inline src/OmniSharp.Roslyn/MetadataHelper.cs]: no 😧

**Comment by adamralph** [comment]
All kinds of awesome

**Comment by filipw** [comment]
thanks, pushed the mentioned changes - should be good to go now :shipit:

**Comment by DustinCampbell** [comment]
:shipit:

**Pull request merged**